server/lib/nl/utils.py

```diff
@@ -167,6 +170,116 @@ def ranked_svs_for_place(place: str, svs: List[str],
   return [sv for sv, _ in svs_with_vals]
 
 
+# Given a place and a list of existing SVs, this API ranks the SVs
+# per the growth rate of the time-series.
+def rank_svs_by_growth_rate(place: str, svs: List[str],
+                            growth_direction: detection.TimeDeltaType,
+                            rank_order: detection.RankingType) -> List[str]:
+  series_data = util.series_core(entities=[place],
+                                 variables=svs,
+                                 all_facets=False)
+
+  svs_with_vals = []
+  for sv, place_data in series_data['data'].items():
+    if place not in place_data:
+      continue
+    series = place_data[place]['series']
+    if len(series) < 2:
+      continue
+
+    try:
+      net_growth_rate = compute_growth_rate(series)
+    except Exception as e:
+      logging.error('Growth rate computation failed: %s', str(e))
+      continue
+
+    if net_growth_rate > 0 and growth_direction != detection.TimeDeltaType.INCREASE:
+      continue
+    if net_growth_rate < 0 and growth_direction != detection.TimeDeltaType.DECREASE:
+      continue
+
+    svs_with_vals.append((sv, net_growth_rate))
+
+  if not rank_order:
+    reverse = True if growth_direction == detection.TimeDeltaType.INCREASE else False
+  else:
+    reverse = True  # Descending order
+    if growth_direction == detection.TimeDeltaType.INCREASE and rank_order == detection.RankingType.LOW:
+      # lowest growing
+      reverse = False
+    elif growth_direction == detection.TimeDeltaType.DECREASE and rank_order == detection.RankingType.HIGH:
+      # highest shrinking
+      reverse = False
+
+  logging.info(f'Rank order: {reverse}')
+  svs_with_vals = sorted(svs_with_vals,
+                         key=lambda pair: pair[1],
+                         reverse=reverse)
+  logging.info(svs_with_vals)
+  return [sv for sv, _ in svs_with_vals]
+
+
+# Computes net growth-rate for a time-series including only recent (since 2012) observations.
+def compute_growth_rate(series: List[Dict]) -> float:
+  latest = None
+  earliest = None
+  for s in series:
```

Comment:
series should already be sorted. So can directly use the first and last item?

Reply:
Ah interesting.  I didn't find it in https://docs.datacommons.org/api/rest/v1/bulk/observations/series, but could have missed.

But still need to loop to handle _THRESHOLD_DATE_FOR_GROWTH_RATE, so I'll just keep this for now.

Reply:
I've added TODO to revisit.

Reply:
Good point, this needs to be added in the documentation. This was one of the improvement in v1 API